### PR TITLE
added whois go plugin

### PR DESCRIPTION
**Pull request by yasharne**
Go plugin to get the remaining time until the domain expiry

### CODE REVIEW

File: modules/whoisquery/provider.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+}
+
+func newProvider(config Config) (provider, error) {
+	sourceDomain := string(config.Source)
+
+	return &fromNet{domainAddress: sourceDomain}, nil
+}
+
+func (f fromNet) remainingTime() (float64, error) {
+
```
Comment (ilyam8) [inline modules/whoisquery/provider.go:27]: i suggest to remove empty line

File: modules/whoisquery/provider.go [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+	sourceDomain := string(config.Source)
+
+	return &fromNet{domainAddress: sourceDomain}, nil
+}
+
+func (f fromNet) remainingTime() (float64, error) {
+	
+	raw, err := whois.Whois(f.domainAddress)
+
+	if err == nil {
```
Comment (ilyam8) [inline modules/whoisquery/provider.go:30]: make the flow (return early if there is fail scenarion)

``go
if err != nil {
  ...
}

...
```

File: modules/whoisquery/provider.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+				expiryRaw += "T0:00:00Z"
+			}
+			expiry, _ := time.Parse(time.RFC3339, expiryRaw)
+			remainingToExpireSeconds := time.Until(expiry).Seconds()
+			return remainingToExpireSeconds, nil
+		}
+	}
+
+	return -1, fmt.Errorf("%v", err)
+}
```
Comment (ilyam8) [inline modules/whoisquery/provider.go:48]: add new line

File: modules/whoisquery/whoisquery.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+	module.Register("whoisquery", creator)
+}
+
+func New() *WhoisQuery {
+	return &WhoisQuery{
+		Config: Config{
+			Timeout:       web.Duration{Duration: time.Second * 2},
+			DaysUntilWarn: 14,
+			DaysUntilCrit: 7,
```
Comment (ilyam8) [inline modules/whoisquery/whoisquery.go:27]: i think we need to tune params, like 3 month warning, 1 month critical. What do you think?

File: modules/whoisquery/whoisquery.go [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+		},
+	}
+}
+
+type Config struct {
+	web.ClientTLSConfig `yaml:",inline"`
+	Timeout             web.Duration
```
Comment (ilyam8) [inline modules/whoisquery/whoisquery.go:34]: Timeout and web.ClientTLSConfig are not used

the library you are using has hardcoded timeouts

https://github.com/likexian/whois-go/blob/92cfa71d7edfc954b9661b2676edfb360c32eb3c/whois.go#L102

**Comment by ilyam8** [comment]
+ we need tests (good example is [docker_engine](https://github.com/netdata/go.d.plugin/blob/master/modules/docker_engine/docker_engine_test.go))

File: modules/whoisquery/provider.go (hunk 3 above)
Comment (yasharne) [reply modules/whoisquery/provider.go:48]: Can you explain more where should I put a new line?

Comment (ilyam8) [reply modules/whoisquery/provider.go:48]: right after line 48

File: modules/whoisquery/collect.go [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+	return mx, nil
+}
+
+func (x WhoisQuery) collectExpiration(mx map[string]int64, remainingTime float64) {
+	mx["expiry"] = int64(remainingTime)
+	mx["days_until_expiration_warning"] = x.DaysUntilWarn
+	mx["days_until_expiration_critical"] = x.DaysUntilCrit
+
+}
```
Comment (ilyam8) [inline modules/whoisquery/collect.go:29]: new line

Screenshot 2020-04-07 at 23 48 27

File: modules/whoisquery/provider.go [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+	if (!validDomain) {
+		return nil, fmt.Errorf("Incorrect domain pattern: %v", sourceDomain)
+	}
+	return &fromNet{domainAddress: sourceDomain}, nil
+}
+
+func (f fromNet) remainingTime() (float64, error) {
+	raw, err := whois.Whois(f.domainAddress)
+	if err != nil {
+		return -1, fmt.Errorf("%v", err)
```
Comment (ilyam8) [inline modules/whoisquery/provider.go:32]: make it `return 0, err`

File: modules/whoisquery/provider.go [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+	}
+
+	result, parserErr := whoisparser.Parse(raw)
+	if parserErr != nil {
+		return -1, fmt.Errorf("%v", parserErr)
```
Comment (ilyam8) [inline modules/whoisquery/provider.go:37]: same here

File: modules/whoisquery/collect.go [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+
+func (x *WhoisQuery) collect() (map[string]int64, error) {
+	remainingTime, err := x.prov.remainingTime()
+	if err != nil {
+		return nil, err
+	}
+
+	if remainingTime == -1 {
+		return nil, fmt.Errorf("no domain was provided by '%s'", x.Config.Source)
+	}
```
Comment (ilyam8) [inline modules/whoisquery/collect.go:15]: i guess `if err != nil` below is enough, isnt it? 

is it possible to get nil error and still not actual value?

File: modules/whoisquery/whoisquery_test.go [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+	remTime float64
+	err   bool
+}
+
+func (m mockProvider) remainingTime() (float64, error) {
+	if m.err {
+		return -1, errors.New("mock remaining time error")
+	}
+	return m.remTime, nil
+}
```
Comment (ilyam8) [inline modules/whoisquery/whoisquery_test.go:129]: new line

File: modules/whoisquery/whoisquery_test.go [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+	assert.False(t, whoisquery.Check())
+}
+
+func TestWhoisQuery_Collect(t *testing.T) {
+	whoisquery := New()
+	whoisquery.prov = &mockProvider{remTime: 12345.678}
+
+	collected := whoisquery.Collect()
+
+	assert.NotZero(t, collected)
```
Comment (ilyam8) [inline modules/whoisquery/whoisquery_test.go:94]: `assert.NotZero(t, collected)` is not enough

create `expected map[string]int64` and use assert.Equal

File: modules/whoisquery/whoisquery_test.go (hunk 10 above)
Comment (yasharne) [reply modules/whoisquery/whoisquery_test.go:129]: Do I need to change `-1` to `0` in the line 126 as well?

Comment (ilyam8) [reply modules/whoisquery/whoisquery_test.go:129]: i meant line 126, if there is an error you return the error + zero values. There is no need to indicate that value is wrong using -1

**Comment by ilyam8** [comment]
@yasharne dont forget about main.go

**Comment by ilyam8** [comment]
+ README

**Comment by yasharne** [comment]
@ilyam8 I have updated the README and main.go in the first commit

**Comment by ilyam8** [comment]
somehow i missed main.go update

as for README i mean module readme, there is not README.md in the module dir

**Comment by ilyam8** [comment]
@yasharne there was merge conflict, i fixed it. 

@joelhans take a look at the docs. For some reason cant assign you as a reviewver.

File: modules/whoisquery/README.md [hunk 12]
```diff
@@ -0,0 +1,41 @@
+# Whois domain expiry monitoring with Netdata
+This module checks the remaining time until a domain is expired
```
Comment (joelhans) [inline modules/whoisquery/README.md:2]:
```suggestion

This collector module checks the remaining time until a domain is expired.
```

File: modules/whoisquery/README.md [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+
+## Charts
+It produces the following chart:
```
Comment (joelhans) [inline modules/whoisquery/README.md:5]:
```suggestion
This collector produces the following chart:
```

File: modules/whoisquery/README.md [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+  - name   : my_site
+    source : my_site.com
+    
+  - name   : my_another_site
+    source : my_another_site.com
+    days_until_expiration_critical: 20
+
+```
+
+For all available options and defaults please see module [configuration file](https://github.com/netdata/go.d.plugin/blob/master/config/go.d/whoisquery.conf).
```
Comment (joelhans) [inline modules/whoisquery/README.md:35]:
```suggestion
For all available options and defaults please, see the module's [configuration file](https://github.com/netdata/go.d.plugin/blob/master/config/go.d/whoisquery.conf).
```

File: modules/whoisquery/README.md [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+Edit the `go.d/whoisquery.conf` configuration file using `edit-config` from the your agent's [config
+directory](../../../../docs/step-by-step/step-04.md#find-your-netdataconf-file), which is typically at `/etc/netdata`.
+
+```bash
+cd /etc/netdata # Replace this path with your Netdata config directory
+sudo ./edit-config go.d/whoisquery.conf
+```
+Needs only `source`.
+
+Use `days_until_expiration_warning` and `days_until_expiration_critical` for each job to indicate the expiry warning and critical days, the default values are 90 and 30 days.
```
Comment (joelhans) [inline modules/whoisquery/README.md:18]:
```suggestion
Use `days_until_expiration_warning` and `days_until_expiration_critical` for each job to indicate the expiry warning and critical days. The default values are 90 for warning, and 30 days for critical.
```

**Comment by ilyam8** [comment]
thx @yasharne 👍

**Comment by ilyam8** [comment]
@yasharne btw it would be nice if you add whoisquery alarms to the netdata/netdata.

**Comment by yasharne** [comment]
Sure @ilyam8 I've teste the alarms and will add them soon

**Comment by ilyam8** [comment]
i suggest to use [x509 alarms](https://github.com/netdata/netdata/blob/master/health/health.d/x509check.conf) as a base.

**Comment by yasharne** [comment]
@ilyam8 As a matter of fact I did :)

**Pull request merged**